### PR TITLE
Adds com.github.virtual_puppet_project.vpuppr

### PR DESCRIPTION
**Pull request by orowith2os**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)* **This app requires access to all the user's home directory, since it uses its own file chooser at the moment. As a workaround, this setting has been set to read-only, so it's less of a security hole.**
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type: **There is a file that contains all of the licenses that the project uses, so information about those can be included.**
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:** https://github.com/virtual-puppet-project/vpuppr I've talked with the creator of the project and gotten permission to distribute it as a flatpak.
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: com.github.you_win.VPupPr.appdata.xml
Comment (JakobDev) [inline com.github.you_win.VPupPr.appdata.xml:10]:
```suggestion
    <release type="stable" version="0.8.1"></release>
```

File: com.github.you_win.VPupPr.appdata.xml
Comment (JakobDev) [inline com.github.you_win.VPupPr.appdata.xml:29]: This is not valid. See the [specification](https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-screenshots).

**Review by JakobDev — COMMENTED** [review]
Just a Tip: If you struggle with the Appdata file give [jdAppdataEdit](https://flathub.org/apps/details/com.gitlab.JakobDev.jdAppdataEdit) at try.

File: com.github.you_win.VPupPr.appdata.xml
Comment (orowith2os) [reply com.github.you_win.VPupPr.appdata.xml:29]: I am very well aware of this. I plan on properly putting the screenshots there if/when the project is added to Flathub, just so that there's less hassle of having to deal with online image links and whatnot.

Comment (JakobDev) [reply com.github.you_win.VPupPr.appdata.xml:29]: That will not work. The AppStream file is validated on each build. If your AppStream is not valid, the App can't be added to Flathub. The workaround here is to use the URL of the Screenshots of your current working Repo and chance the URL after your App has been added to Flathub and you have your own Flathub Repo.

File: com.github.you_win.VPupPr.appdata.xml
Comment (orowith2os) [reply com.github.you_win.VPupPr.appdata.xml]: There we go. I've added the github repo using the tool you linked (thanks for that, btw!) and it added all of the releases automatically.

File: com.github.you_win.VPupPr.appdata.xml
Comment (JakobDev) [reply com.github.you_win.VPupPr.appdata.xml]: Yes

Comment (orowith2os) [reply com.github.you_win.VPupPr.appdata.xml]: I've added one, as the other ended up crashing the app and I don't know if that would affect anything else. It should be ready to go now, I think?

File: com.github.you_win.VPupPr.metainfo.xml
Comment (hfiguiere) [inline com.github.you_win.VPupPr.metainfo.xml:4]: So I will repeat what I said. This does not match the project URL.

You download from `https://github.com/virtual-puppet-project/vpuppr`
and this is also what is in the URL fields in the appstream file.

File: com.github.you_win.VPupPr.yml
Comment (hfiguiere) [inline com.github.you_win.VPupPr.yml:12]: that is a lot of filesystem permissions.

File: com.github.you_win.VPupPr.yml
Comment (hfiguiere) [inline com.github.you_win.VPupPr.yml:9]: if I read the comment right, you want this at build time?

1. this is not what this does
2. build is offline so it wouldn't work anyway.

The question is do you need network access at runtime?

File: com.github.you_win.VPupPr.yml
Comment (hfiguiere) [inline com.github.you_win.VPupPr.yml:22]: for the record, only 128x128 is required. So I would reduce the number of icons here, and if you want that many, make sure upstream provide them instead.

**Review by hfiguiere — CHANGES_REQUESTED** [review]
Have you submitted the extra files upstream (.destop, .appdata.xml, screenshot and icons) ?

File: com.github.you_win.VPupPr.yml
Comment (orowith2os) [reply com.github.you_win.VPupPr.yml:12]: I've limited it to what I can, so as to not break the user experience and have to tinker with the flatpak permissions too much. I've set the home folder to read-only, so it's not as dangerous, I hope.

**Comment by orowith2os** [comment]
I've submitted the icons upstream, I don't feel that the rest should be included. I may add the screenshots soon.

File: com.github.you_win.VPupPr.yml
Comment (orowith2os) [reply com.github.you_win.VPupPr.yml:22]: I've submitted the icons to upstream already, so I'd say this suggestion is already covered. I'll consider lowering it to 128x128, but I don't see any reason to.

File: com.github.you_win.VPupPr.metainfo.xml
Comment (hfiguiere) [reply com.github.you_win.VPupPr.metainfo.xml:4]: I said it **didn't match the project**. There was NOT A SINGLE REFERENCE to 'youwin' beside the app-id.

File: com.github.you_win.VPupPr.yml
Comment (hfiguiere) [reply com.github.you_win.VPupPr.yml:12]: why do you need to read the home directory?

File: com.github.you_win.VPupPr.metainfo.xml
Comment (orowith2os) [reply com.github.you_win.VPupPr.metainfo.xml:4]: I thought the project was still available at the you-win/vpuppr link, so I put that. I've fixed it, and will push the commits soon.

File: com.github.you_win.VPupPr.yml
Comment (orowith2os) [reply com.github.you_win.VPupPr.yml]: I've removed them and just used `128`, if I get any complaints about it looking blurry, I will add back `256`.

File: com.github.you_win.VPupPr.yml
Comment (orowith2os) [reply com.github.you_win.VPupPr.yml]: They go to ~/.var/app/com.github.you_win.VPupPr/data/godot/app_userdata/OpenSeeFaceGD/venv/ for the face tracking. Network access is also required if the user wants to use an external instance of the face tracker.

Comment (orowith2os) [reply com.github.you_win.VPupPr.yml]: The dependencies now get installed to ~/.var/app/com.github.virtual_puppet_project.vpuppr/data/godot/app_userdata/OpenSeeFaceGD/venv/

**Comment by hfiguiere** [comment]
> no .desktop file is required for upstream 

It's a bug. It is the standard mechanism to provide a menu entry on Linux. It has been since *last century*. (that's more than two decades)

**Comment by hfiguiere** [comment]
bot, build com.github.virtual_puppet_project.vpuppr

File: com.github.virtual_puppet_project.vpuppr.yml
Comment (hfiguiere) [inline com.github.virtual_puppet_project.vpuppr.yml:43]:
```suggestion
```

and remove the file from the repository.

File: com.github.virtual_puppet_project.vpuppr.yml
Comment (hfiguiere) [inline com.github.virtual_puppet_project.vpuppr.yml:22]:
```suggestion
      - install -Dm 644 com.github.virtual_puppet_project.vpuppr-128.png /app/share/icons/hicolor/128x128/apps/com.github.virtual_puppet_project.vpuppr.png
      - install -Dm 644 com.github.virtual_puppet_project.vpuppr-256.png /app/share/icons/hicolor/256x256/apps/com.github.virtual_puppet_project.vpuppr.png
```

1. install create the dir. 
2. you were trying to create it after.

as I said I'd keep the 256.

File: com.github.virtual_puppet_project.vpuppr.yml
Comment (hfiguiere) [inline com.github.virtual_puppet_project.vpuppr.yml:47]: I would have kept this one (I was really taling about having the < 128px here)

```suggestion
      - type: file
        path: icons/com.github.virtual_puppet_project.vpuppr-256.png
```

File: com.github.you_win.VPupPr.oldappdata
Comment (hfiguiere) [inline com.github.you_win.VPupPr.oldappdata:1]: this file should be removed.

File: com.github.virtual_puppet_project.vpuppr.yml
Comment (hfiguiere) [inline com.github.virtual_puppet_project.vpuppr.yml:26]: the appdata file is still NOT installed. and it should be listed in the sources too.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: com.github.virtual_puppet_project.vpuppr.yml
Comment (orowith2os) [reply com.github.virtual_puppet_project.vpuppr.yml:26]: Do you mean `com.github.virtual_puppet_project.vpuppr.metainfo.xml`? If you do, how would I go about installing it?

File: com.github.virtual_puppet_project.vpuppr.yml
Comment (hfiguiere) [reply com.github.virtual_puppet_project.vpuppr.yml:9]: So in a non standard location that is readonly. I guess this was tested, right?

File: com.github.virtual_puppet_project.vpuppr.yml
Comment (hfiguiere) [reply com.github.virtual_puppet_project.vpuppr.yml:26]: https://github.com/flathub/flathub/wiki/AppData-Guidelines#path-and-filename

File: com.github.virtual_puppet_project.vpuppr.yml
Comment (hfiguiere) [reply com.github.virtual_puppet_project.vpuppr.yml:43]: this was not resolved. the commented line ought to be removed too.

File: com.github.virtual_puppet_project.vpuppr.yml
Comment (orowith2os) [reply com.github.virtual_puppet_project.vpuppr.yml:9]: It's stored inside of the flatpak configuration directory that flatpak gives it, it's not an issue, as the flatpak has full read and write access to its own configuration directories. 
Flatpak gives this flatpak full access to ~/.var/app/com.github.virtual_puppet_project.vpuppr/, and that's all it NEEDS.

Comment (hfiguiere) [reply com.github.virtual_puppet_project.vpuppr.yml:9]: Note that there is no pip in the runtime. Only the SDK. Not even sure the venv module is there.

Comment (orowith2os) [reply com.github.virtual_puppet_project.vpuppr.yml:9]: Ah yeah, that's the directory I meant. 
And, it seems like pip works just fine inside of the flatpak environment? if I disable complete access to the home folder, it'll still install just fine. I'll do some more testing to be sure. But it seems to work, at least.

File: com.github.virtual_puppet_project.vpuppr.yml
Comment (orowith2os) [reply com.github.virtual_puppet_project.vpuppr.yml:30]: I've added the file, with steps to add it to the appropriate directories. However, it says the .xml file is not valid, despite the validation instructions you linked saying it is.

File: com.github.virtual_puppet_project.vpuppr.yml
Comment (hfiguiere) [inline com.github.virtual_puppet_project.vpuppr.yml:22]:
```suggestion
       - install -Dm644 com.github.virtual_puppet_project.vpuppr.metainfo.xml -t /app/share/metainfo/
```

**Review by hfiguiere — CHANGES_REQUESTED** [review]
if you don't tell the error, I can't help.

**Comment by orowith2os** [comment]
> if you don't tell the error, I can't help.

image

**Comment by hfiguiere** [comment]
btw we are still waiting for the bot to start a build.

**Comment by orowith2os** [comment]
> ah yeah that one is actually a bug with `flatpak-builder` (a recent regression, where something that was ignored is now an error and it shouldn't).
> 
> It's recommended to use the flatpak version that can be installed using `flatpak install org.flatpak.Builder`. And replace `flatpak-builder` by `flatpak run org.flatpak.Builder` when calling it.

ah, got it. I've aliased flatpak-builder to the command, so it'll work just fine now. No building errors anymore over here! I think we're good to go, what say ye?

**Comment by orowith2os** [comment]
Turns out the app is broken when loading models with a read-only filesystem. I've set it to the default (--filesystem=home) so that it works.

**Comment by orowith2os** [comment]
The app is working properly. It does not seem to require any more changes. Any problems it currently has (not using the file chooser portal) will be worked out on the next release of vpuppr.

**Comment by JakobDev** [comment]
bot, build com.github.virtual_puppet_project.vpuppr

**Comment by hfiguiere** [comment]
> It does not seem to require any more changes. Any problems it currently has (not using the file chooser portal) will be worked out on the next release of vpuppr.

Also aarch64 support. (building from source?)

**Comment by orowith2os** [comment]
> > It does not seem to require any more changes. Any problems it currently has (not using the file chooser portal) will be worked out on the next release of vpuppr.
> 
> Also aarch64 support. (building from source?)

Aarch64 support seems to mainly rely on onnxruntime running on ARM64 devices. If that requirement is met, I can see about remaking the manifest to build from source instead. However, currently, it relies on modified Godot versions, and is undergoing a rewrite. Maybe, with the release of Godot 4 (with OpenGL support) it can run on Aarch64.

**Comment by barthalion** [comment]
@orowith2os Who from the upstream team should be given write access to the Flathub repo?

**Comment by orowith2os** [comment]
@you-win and @ItsRogueRen are the ones I'm comfortable giving write access, I'll leave the final decision up to you-win on who to give write access.

**Comment by ItsRogueRen** [comment]
You-win should be the one, it's his protect so he gets final say on everything

**Comment by you-win** [comment]
Myself and @orowith2os should have write access. @ItsRogueRen isn't really on the programming side, so access for them isn't necessary.

**Pull request closed without merging**
